Merge request for customer

YurkinED opened the pull request:
Реализация финального задания спринта 13
[1]Функциональность  «Отзывы»  (+)
[2]Функциональность «Поиск»
[3]Функциональность «Общие фильмы»
[4]Функциональность «Рекомендации»
[5]Функциональность «Лента событий»
[6]Удаление фильмов и пользователей
[7]Добавление режиссёров в фильмы
[8]Вывод самых популярных фильмов по жанру и годам

YurkinED committed:
Update FilmorateApplication.java

IvanSuchilin committed:
first changes

YurkinED committed:
Update FilmorateApplication.java

YurkinED committed:
Добавление данных по задаче 3

Добавление контроллера /films/common?userId={userId}&friendId={friendId}

Solntmore committed:
<feat> В объект film добавлен Set, хранящий Id режиссеров

<feat> В БД добавлена таблица directors, содержащая id и имя режиссеров

<feat> Добавлена модель Director, содержащая в себе данные о режиссере

<feat> Метод `installFilmGenres` переименован в `installFilmGenresAndDirectors` с добавленным функционалом добавления и обновления списка режиссеров

<feat> Добавлен контроллер `DirectorController` обрабатывающий запросы по режиссерам

<feat> Добавлен сервис `DirectorService` обрабатывающий запросы по режиссерам

<feat> Добавлен storage `DirectorDbStorage` выполняющий запросы в БД

<refactor> Внесены изменения в порядок работы с жанрами без изменения логики. Исправление, слой `GenreController` теперь обращается к `GenreService, `а не напрямую в `GenreDbStorage`

<feat> В класс `FilmController` добавлен обработчик запроса фильмов режиссера

<feat> В класс `FilmService` и `FilmDbStorage` добавлены методы, реализующие логику подбора и сортировки фильмов по id режиссера

<feat> Добавлен класс `DirectorDbStorageTest` с тестами на новый функционал для режиссеров

<refactor> В класс `FilmDbStorage` добавлен тест, проверяющий корректность работы поиска фильмов по режиссеру.

YurkinED committed:
Добавление функционала поиска наиболее популярных фильмов

YurkinED committed:
Update SqlQueryConstantsForFilm.java

Solntmore committed:
<feat> В объект film добавлен Set, хранящий Id режиссеров

<feat> В БД добавлена таблица directors, содержащая id и имя режиссеров

<feat> Добавлена модель Director, содержащая в себе данные о режиссере

<feat> Метод `installFilmGenres` переименован в `installFilmGenresAndDirectors` с добавленным функционалом добавления и обновления списка режиссеров

<feat> Добавлен контроллер `DirectorController` обрабатывающий запросы по режиссерам

<feat> Добавлен сервис `DirectorService` обрабатывающий запросы по режиссерам

<feat> Добавлен storage `DirectorDbStorage` выполняющий запросы в БД

<refactor> Внесены изменения в порядок работы с жанрами без изменения логики. Исправление, слой `GenreController` теперь обращается к `GenreService, `а не напрямую в `GenreDbStorage`

<feat> В класс `FilmController` добавлен обработчик запроса фильмов режиссера

<feat> В класс `FilmService` и `FilmDbStorage` добавлены методы, реализующие логику подбора и сортировки фильмов по id режиссера

<feat> Добавлен класс `DirectorDbStorageTest` с тестами на новый функционал для режиссеров

<refactor> В класс `FilmDbStorage` добавлен тест, проверяющий корректность работы поиска фильмов по режиссеру.

Solntmore committed:
<feat> В объект film добавлен Set, хранящий Id режиссеров

<feat> В БД добавлена таблица directors, содержащая id и имя режиссеров

<feat> Добавлена модель Director, содержащая в себе данные о режиссере

<feat> Метод `installFilmGenres` переименован в `installFilmGenresAndDirectors` с добавленным функционалом добавления и обновления списка режиссеров

<feat> Добавлен контроллер `DirectorController` обрабатывающий запросы по режиссерам

<feat> Добавлен сервис `DirectorService` обрабатывающий запросы по режиссерам

<feat> Добавлен storage `DirectorDbStorage` выполняющий запросы в БД

<refactor> Внесены изменения в порядок работы с жанрами без изменения логики. Исправление, слой `GenreController` теперь обращается к `GenreService, `а не напрямую в `GenreDbStorage`

<feat> В класс `FilmController` добавлен обработчик запроса фильмов режиссера

<feat> В класс `FilmService` и `FilmDbStorage` добавлены методы, реализующие логику подбора и сортировки фильмов по id режиссера

<feat> Добавлен класс `DirectorDbStorageTest` с тестами на новый функционал для режиссеров

<refactor> В класс `FilmDbStorage` добавлен тест, проверяющий корректность работы поиска фильмов по режиссеру.

IvanSuchilin committed:
added delete user endpoint

IvanSuchilin committed:
add film deleting and tests

MikhailKuznets committed:
feat: создать рабочую ветку add-recommendations

IvanSuchilin committed:
add feed table

IvanSuchilin committed:
fixed with pr

YurkinED committed:
Форматирование SQL запроса

YurkinED committed:
Поднял уровень логирования

YurkinED committed:
Убрал неиспользуемый импорт

Solntmore committed:
<refactor> В тестах добавлен `@DisplayName`

<refactor> Во всех константах `SqlQueryConstantsForFilm` и `SqlQueryConstantsForUser` символ * заменен на конкретные столбцы

<refactor> Вся логика сортировки перенесена из кода в SQL запросы

<feat> В объект film добавлен Set, хранящий Id режиссеров

<feat> В БД добавлена таблица directors, содержащая id и имя режиссеров

<feat> Добавлена модель Director, содержащая в себе данные о режиссере

<feat> Метод `installFilmGenres` переименован в `installFilmGenresAndDirectors` с добавленным функционалом добавления и обновления списка режиссеров

<feat> Добавлен контроллер `DirectorController` обрабатывающий запросы по режиссерам

<feat> Добавлен сервис `DirectorService` обрабатывающий запросы по режиссерам

<feat> Добавлен storage `DirectorDbStorage` выполняющий запросы в БД

<refactor> Внесены изменения в порядок работы с жанрами без изменения логики. Исправление, слой `GenreController` теперь обращается к `GenreService, `а не напрямую в `GenreDbStorage`

<feat> В класс `FilmController` добавлен обработчик запроса фильмов режиссера

<feat> В класс `FilmService` и `FilmDbStorage` добавлены методы, реализующие логику подбора и сортировки фильмов по id режиссера

<feat> Добавлен класс `DirectorDbStorageTest` с тестами на новый функционал для режиссеров

<refactor> В класс `FilmDbStorage` добавлен тест, проверяющий корректность работы поиска фильмов по режиссеру.

IvanSuchilin committed:
add showUsersFeeds

YurkinED committed:
Исправление пробелов

YurkinED committed:
Merge pull request #11 from YurkinED/add-remove-endpoint

Add remove endpoint

YurkinED committed:
Merge branch 'develop' into add-common-films

YurkinED committed:
BUg fix when merge

YurkinED committed:
Merge pull request #7 from YurkinED/add-common-films

Add common films

MikhailKuznets committed:
feat: создать sql-запрос add-recommendations

MikhailKuznets committed:
feat: добавить функционал рекоммендаций фильмов

YurkinED committed:
Merge branch 'develop' into add-director

YurkinED committed:
Resolve conflicts

YurkinED committed:
Merge pull request #12 from YurkinED/add-director

Add director

mkudriavtsev committed:
feat: add reviews

MikhailKuznets committed:
refactor: попытка смержить с develop

mkudriavtsev committed:
fix: fixed minor typos

MikhailKuznets committed:
refactor: смержил с develop, добился корректного прохождения тестов

mkudriavtsev committed:
fix: add logging

mkudriavtsev committed:
Merge branch 'develop' into add-reviews

YurkinED committed:
Update schema.sql

MikhailKuznets committed:
refactor: оптимизировал метод getRocommendations

IvanSuchilin committed:
add creating feed for friends and likes

mkudriavtsev committed:
Merge pull request #14 from YurkinED/add-reviews

Добавлена функциональность «Отзывы»

YurkinED committed:
Bug fix

Solntmore committed:
<feat> Добавлена возможность выводить топ-N фильмов по количеству лайков.
<feat> Добавлены константы выполняющие поиск и сортировку по различным параметрам.
<refactor> Переработан контролер showMostLikedFilms, поддерживает GET /films/popular?count={limit}&genreId={genreId}&year={year}.

IvanSuchilin committed:
Merge branch 'develop' into add-feed

IvanSuchilin committed:
after merging with develope

YurkinED committed:
Merge branch 'develop' into add-recommendations

YurkinED committed:
Merge pull request #13 from YurkinED/add-recommendations

Add recommendations

IvanSuchilin committed:
add feed into review

IvanSuchilin committed:
add review feeds

IvanSuchilin committed:
add fixed

mkudriavtsev committed:
feat: start to implement

YurkinED committed:
Обновление схемы

YurkinED committed:
Обновление поля режисеров для фильмов

IvanSuchilin committed:
add check to update review'

mkudriavtsev committed:
fix: fix to postman tests

mkudriavtsev committed:
Merge remote-tracking branch 'origin/add-director' into add-search

IvanSuchilin committed:
update review upd

mkudriavtsev committed:
feat: continue

mkudriavtsev committed:
fix: fix updateReview method

IvanSuchilin committed:
done all tests

mkudriavtsev committed:
feat: search functional implemented

mkudriavtsev committed:
Merge remote-tracking branch 'origin/develop' into add-reviews

mkudriavtsev committed:
fix: fix minor typos

mkudriavtsev committed:
fix: fix minor typos

IvanSuchilin committed:
after merging with newReview

mkudriavtsev committed:
Merge pull request #19 from YurkinED/add-search

Функциональность «Поиск»

mkudriavtsev committed:
Merge pull request #18 from YurkinED/add-reviews

Внесены правки для прохождения тестов Postman

IvanSuchilin committed:
Merge branch 'develop' into add-feed

YurkinED committed:
Изменение арзитектуры

YurkinED committed:
Remove unised method

IvanSuchilin committed:
after merging with actual dev

YurkinED committed:
Update FilmController.java

YurkinED committed:
Update FilmDbStorage.java

YurkinED committed:
Update FilmDbStorage.java

YurkinED committed:
Update FilmDbStorage.java

IvanSuchilin committed:
add constants for feedmethod

YurkinED committed:
Merge pull request #21 from YurkinED/add-feed

Add feed

YurkinED committed:
Merge branch 'develop' into add-most-populars

YurkinED committed:
Update FilmDbStorage.java

YurkinED committed:
Update SqlQueryConstantsForFilm.java

YurkinED committed:
Update FilmController.java

YurkinED committed:
Merge pull request #15 from YurkinED/add-most-populars

Add most populars

YurkinED committed:
Final review

YurkinED committed:
Update README.md

mkudriavtsev reviewed (approved):
Отлично потрудились:)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -30,7 +31,11 @@ public ErrorResponse handleSystemExceptions(final NullPointerException e) {
 
```
В класс следует добавить логирование) Для этого достаточно использовать аннотацию из lombok ->> [@slf4j](https://github.com/slf4j). Далее в каждый наш хендлер следует добавить логирование сообщения об ошибке)
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем Throwable, а возвращать статус 500 INTERNAL_SERVER_ERROR

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -32,22 +33,41 @@ public Collection<Film> findAllFilms() {
+
+
+    @GetMapping("director/{directorId}")
+    public List<Film> showDirectorFilms(@PathVariable int directorId, @RequestParam(name = "sortBy") String sortBy) {
```
Можно было бы добавить Enum с константами для видов сортировки и принимать его, а не строку) Таким образом еще на уровне фильтров спринга произойдет выброс ошибки в случае, если будет передан некорректных вид)

IvanSuchilin replied:
добавлен enum SortingChoices

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -72,5 +92,17 @@ public void removeLikeFromFilm(@PathVariable int filmId, @PathVariable int userI
+    }
 
+    @GetMapping("/search")
+    public List<Film> searchFilms(@RequestParam String query, @RequestParam List<String> by) {
```
Здесь так же можно было бы использовать Enum для выбора параметров для поиска)

IvanSuchilin replied:
добавлен enum SearchingParts

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -32,22 +33,41 @@ public Collection<Film> findAllFilms() {
-            log.debug("Получен запрос GET /films/popular?count={}. Показать топ {} фильмов по лайкам.", count, count);
-            return filmService.showMostLikedFilms(count.get());
+    @GetMapping(value = {"/popular"})
+    public List<Film> showMostLikedFilms(@RequestParam(name = "count", required = false, defaultValue = "10") Integer limit,
```
Было бы классно использовать для значений аргументов аннотации валидации) Например, limit не может быть отрицательным - следует использовать `@...`
Так же как и для года можно было бы ввести ограничение, так как еще на первом спринте этого проекта было введено ограничение по году)
Над классом контроллера необоходимо навесить аннотацию `@Validated`, таким образом перед передачей значения в метод, оно будет проверяться на соответствие, если оно не будет удовлетворять условию, будет выброшена ошибка валидации и обработана, а статус вернется красивым(400 BAD_REQUEST), а для красивого ответа от сервера нужно использовать еще один обработчик
```
@ExceptionHandler
public ResponseEntity< String > exc(ConstraintViolationException ex){
      return new ResponseEntity<>(ex.getMessage(), HttpStatus.BAD_REQUEST);
}
```

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @GetMapping
+    List<Review> findAllOrByFilmIdReviews(
```
В данном случае хорошо было бы так же использовать аннотации валидации для значений аргументов)

mkudriavtsev replied:
Готово

TyutterinYakov replied:
Но они отсутствуют)
Например, для count логично было бы использовать `@Positive`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,17 @@
+
+@Data
+@AllArgsConstructor
+public class Review {
```
В данном случае удобно было бы использовать валидацию по группам)

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,17 @@
+@AllArgsConstructor
+public class Review {
+    private Integer reviewId;
+    private String content;
```
Например, content должен быть обязательно передан при создании и не может состоять из пробелов или быть пустым)
`@NotBlank`

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,17 @@
+public class Review {
+    private Integer reviewId;
+    private String content;
+    @JsonProperty("isPositive")
```
Значение этого поля может иметь лишь два состояния, следовательно стоит использовать `@NotNull`

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,17 @@
+    @JsonProperty("isPositive")
+    private Boolean isPositive;
+    private Integer useful;
+    private Integer userId;
```
Идентификатор пользователя и идентификатор фильма должны быть переданы в обязательном порядке, следует использовать `@NotNull`

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,17 @@
+@Data
+@AllArgsConstructor
+public class Review {
+    private Integer reviewId;
```
Идентификатор отзыва должен быть передан при обновлении - `@NotNull(groups = {Update.class})`, поэтому здесь мы указываем группу)

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,83 @@
+    private final ReviewService reviewService;
+
+    @PostMapping
+    Review createReview(@RequestBody Review review) {
```
Не забываем здесь использовать аннотацию @Valid перед аргументом метода, чтобы аннотации валидации стали учитываться)
Кстати, вместо @Valid можно использовать такую аннотацию @Validated(Create.class). Таким образом мы указываем, какую группу аннотаций следует учитывать. Create - это интерфейс, который ты создаешь самостоятельно, он будет пустым и служит лишь для указания группы) Аналогично следует поступить при обновлении, но там будет аннотация @Validated(Update.class)
Соответствующие группы следует указывать в аннотациях валидации в классе-модели)

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/BaseEntity.java:
```diff
@@ -4,9 +4,8 @@
 @EqualsAndHashCode
-@ToString
 public class BaseEntity {
     private int id;
```
Понимаю, что вы зашли уже достаточно далеко, поэтому оставлю как предложение, но для хранения идентификатора лучше использовать тип long, так как типа int может не хватить

MikhailKuznets replied:
Изменил Data Type c int на long

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDBStorage.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    public Review updateUsefulInReview(Review review) {
```
Метод можно было бы сделать приватным и вызывать при удалении/добавлении лайка. Запрос будет выглядеть примерно таким образом)
`update REVIEWS r set useful = (select sum(l.useful) from REVIEW_LIKES l where l.review_id = r.review_id)  where review_id = ?`
Так мы избавимся от лишней логики в сервисе и расчетов в коде
Но хранить в таблице reviews_likes поле useful следует в виде int)

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDBStorage.java:
```diff
@@ -0,0 +1,117 @@
+import static ru.yandex.practicum.filmorate.constants.SqlQueryConstantsForReview.*;
+
+
+@Component
```
Более семантически явно было бы использовать аннотацию `@Repository`, хоть и разницы с `@Component` нет)

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Feed.java:
```diff
@@ -0,0 +1,19 @@
+    private int eventId;
+    private int userId;
+    private int entityId;
+    private String eventType;
```
(предложение) eventType и operation следует представить в виде Enum)
Не думаю, что ради хранения нескольких значений стоит использовать отдельные таблицы в БД)

IvanSuchilin replied:
Класс Feed - добавлены enum-ы (вложенные так как использоются только в Feed)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/DirectorService.java:
```diff
@@ -0,0 +1,41 @@
+        return directorDbStorage.findAllDirectors();
+    }
+
+    public Optional<Director> findDirectorById(@PathVariable int genreId) {
```
Аннотации `@PathVariable` здесь быть не должно)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/DirectorService.java:
```diff
@@ -0,0 +1,41 @@
+
+@Service
+public class DirectorService {
+    private final DirectorDbStorage directorDbStorage;
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,21 +1,21 @@
+
 @Slf4j
 @Service
 public class FilmService {
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/GenreService.java:
```diff
@@ -1,11 +1,12 @@
+import java.util.Optional;
 
 @Service
 public class GenreService {
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/MpaService.java:
```diff
@@ -1,4 +1,4 @@
+package ru.yandex.practicum.filmorate.service;
 
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,113 @@
+public class ReviewService {
+    private final ReviewStorage reviewStorage;
+    private final ReviewValidator reviewValidator;
+    private final FilmDbStorage filmStorage;
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации)

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,15 +1,18 @@
+
 @Slf4j
 @Service
 public class UserService {
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации)

MikhailKuznets replied:
Исправлено

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -5,56 +5,136 @@ DROP TABLE IF EXISTS films_genres CASCADE;
 );
 
-CREATE TABLE likes
+create table likes
```
Первичный ключ в связующей таблице должен быть представлен в виде отдельного столбца, либо же должен быть составным)

YurkinED replied:
ok

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -5,56 +5,136 @@ DROP TABLE IF EXISTS films_genres CASCADE;
+            on delete cascade
+);
+
+create table friends
```
Первичный ключ в связующей таблице должен быть представлен в виде отдельного столбца, либо же должен быть составным)

YurkinED replied:
ok

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -5,56 +5,136 @@ DROP TABLE IF EXISTS films_genres CASCADE;
+    film_id    int REFERENCES films (film_id) ON DELETE CASCADE
+);
+
+CREATE TABLE reviews_likes
```
Первичный ключ в связующей таблице должен быть представлен в виде отдельного столбца, либо же должен быть составным)

YurkinED replied:
ok

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -5,56 +5,136 @@ DROP TABLE IF EXISTS films_genres CASCADE;
+    director_name varchar(60)
+);
+
+CREATE TABLE films_directors
```
Первичный ключ в связующей таблице должен быть представлен в виде отдельного столбца, либо же должен быть составным)

YurkinED replied:
ok

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -5,56 +5,136 @@ DROP TABLE IF EXISTS films_genres CASCADE;
 );
 
-CREATE TABLE films_genres
+create table films_genres
```
Первичный ключ в связующей таблице должен быть представлен в виде отдельного столбца, либо же должен быть составным)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -36,6 +36,8 @@ public void addLikeToFilm(int filmId, int userId) {
                 && filmStorage.findFilmById(filmId).isPresent()) {
```
Для извлечения значения из Optional было бы удобно использовать `orElseThrow(() -> ...)`
Если Optional будет пуст - будет выброшено указанное исключение)

IvanSuchilin replied:
done

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/validators/ReviewValidator.java:
```diff
@@ -0,0 +1,29 @@
+import ru.yandex.practicum.filmorate.model.Review;
+
+import java.util.Objects;
+
```
Класс FilmValidator не отображается, поэтому напишу здесь)

```
package ru.yandex.practicum.filmorate.validators;

import org.springframework.stereotype.Component;
import ru.yandex.practicum.filmorate.exceptions.filmExceptions.InvalidNameException;
import ru.yandex.practicum.filmorate.exceptions.filmExceptions.LongDescriptionException;
import ru.yandex.practicum.filmorate.exceptions.filmExceptions.NegativeDurationException;
import ru.yandex.practicum.filmorate.exceptions.filmExceptions.ReleaseDateException;
import ru.yandex.practicum.filmorate.model.BaseEntity;
import ru.yandex.practicum.filmorate.model.Film;

import java.time.LocalDate;

@Component
public class FilmValidator {
    private final static LocalDate BENCHMARK_FOR_FILM_DATE = LocalDate.of(1895, 12, 28);

    public void validator(Film film) {
        String name = film.getName();
        BaseEntity mpa = film.getMpa();

//Проверка не требуется, так как она реализована с помощью аннотации валидации @NotBlank
        if (name == null || name.isBlank()) {
            throw new InvalidNameException("Нужно заполнить название фильма.");
        }
//Проверка не требуется, так как она реализована с помощью аннотации валидации @Size, но дополнительно следует использовать аннотацию @NotBlank, так как первая допускает null, что может спровоцировать NPE
        if (film.getDescription().length() > 200) {
            throw new LongDescriptionException("Максимальная длина описания — 200 символов.");
        }
        if (film.getReleaseDate().isBefore(BENCHMARK_FOR_FILM_DATE)) {
            throw new ReleaseDateException("Доступная дата релиза не может быть раньше " + BENCHMARK_FOR_FILM_DATE);
        }
//Проверка не требуется, так как она реализована с помощью аннотации @Min(1)
        if (film.getDuration() <= 0) {
            throw new NegativeDurationException("Продолжительность фильма должна быть положительной.");
        }
//Проверка не требуется, так как она реализована с помощью аннотации @NotNull
        if (mpa == null) {
            throw new RuntimeException("Необходимо добавить MPA.");
        }
    }
}
```

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -29,10 +29,12 @@ public class Film {
 
```
Вместо BaseEntity следует использовать Mpa

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -29,10 +29,12 @@ public class Film {
 
     private Set<BaseEntity> genres = new TreeSet<>(Comparator.comparingInt(BaseEntity::getId));
```
Вместо BaseEntity следует использовать Genre

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/validators/ReviewValidator.java:
```diff
@@ -0,0 +1,29 @@
+@Component
+public class ReviewValidator {
+
+    public void validate(Review review) {
```
Теперь этот класс не требуется, так как все проверки будут реализованы с помощью аннотаций)

mkudriavtsev replied:
Готово

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/BaseEntity.java:
```diff
@@ -4,9 +4,8 @@
+@Data
 @EqualsAndHashCode
-@ToString
 public class BaseEntity {
```
Класс User не отображается, поэтому напишу здесь)
```
@Slf4j
@Data
public class User {
    private int id;
//Так же следует использовать `@NotNull`, так как `@Email` допускает null)
    @Email(message = "Email должен быть корректным адресом электронной почты.")
    private String email;
    @NotBlank(message = "Необходимо указать login.")
    @Pattern(regexp = "^\\S*$")
    private String login;
    private String name;
//В данном случае можно было бы использовать аннотацию для проверки даты на то, что она не раньше текущей, но так же необходимо использовать `@NotNull`, так как первая допускает null)
    private LocalDate birthday;
//Раз ты познакомился с интерфейсами и полиморфизмом, может быть стоит теперь использовать интерфейс List, чтобы не привязывать к конкретной его реализации) 
    private ArrayList<Integer> friends = new ArrayList<>();

....
```

MikhailKuznets replied:
Сделано:
Добавлена проверка NotNull к полям: login, email
Добавлена проверка NotNull и PastOrPresent к полю: birthday
Переделан ArrayList -> List

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -186,5 +189,34 @@ private MapSqlParameterSource makeFriendsParameters(int userId, int friendId) {
+                (rs, rowNum) -> makeFeed(id, rs), id);
+    }
+
+    private Feed makeFeed(int id, ResultSet rs) throws SQLException {
```
Хорошо было бы придерживаться порядка во всех классах, чтобы сначала шли публичные методы(предоставляемый API), а уже потом приватные(внутренняя реализация)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -186,5 +189,34 @@ private MapSqlParameterSource makeFriendsParameters(int userId, int friendId) {
+                (rs, rowNum) -> makeFeed(id, rs), id);
+    }
+
+    private Feed makeFeed(int id, ResultSet rs) throws SQLException {
```
Хорошо было бы придерживаться порядка во всех классах, чтобы сначала шли публичные методы(предоставляемый API), а уже потом приватные(внутренняя реализация)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -186,5 +189,34 @@ private MapSqlParameterSource makeFriendsParameters(int userId, int friendId) {
+        return namedParameterJdbcTemplate.getJdbcTemplate().query(SQL_QUERY_SHOW_FEEDS_BY_USER_ID,
+                (rs, rowNum) -> makeFeed(id, rs), id);
+    }
+
```
Хорошо было бы придерживаться порядка во всех классах, чтобы сначала шли публичные методы(предоставляемый API), а уже потом приватные(внутренняя реализация)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -10,11 +9,15 @@
+import java.time.LocalDateTime;
 import java.time.format.DateTimeFormatter;
 import java.util.*;
 import java.util.stream.Collectors;
```
Для соблюдения инкапсуляции поля класса должны быть приватными) Так же `DateTimeFormatter formatter` хорошо было бы сделать `static`, так как независимо от объекта класса значение всегда будет одним)

Проверку 
```
        if (isBlank(name)) {
            user.setName(user.getLogin());
        }
```

Хорошо было бы вынести на уровень контроллера, так как происходит проверка данных, сформированных не в результате работы бизнес логики)


Логику 
```
            User user = new User(
                    userRows.getInt("user_id"),
                    userRows.getString("email"),
                    userRows.getString("login"),
                    userRows.getString("name"),
                    Objects.requireNonNull(userRows.getDate("birthday")).toLocalDate());
```

Следует вынести в отдельный приватный метод, чтобы не дублировать код и для соблюдения Single Responsibility(принципа единственной ответственности)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -186,5 +189,34 @@ private MapSqlParameterSource makeFriendsParameters(int userId, int friendId) {
         return parameters;
     }
 
+    public Collection<Feed> showUsersFeeds(int id) {
```
Работа с событиями должна осуществляться в отдельном DAO, так как это другая сущность в бд)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -10,27 +10,32 @@
 @Component
 public class FilmDbStorage implements FilmStorage {
     private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+    private final DirectorService directorService;
```
Слой нижнего уровня не может зависеть от слоя верхнего уровня. Хорошей архитектурой является та, где каждый слой верхнего уровня зависит от слоя более низкого уровня, но допускает зависимость сервиса от другого сервиса)

YurkinED replied:
ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -165,4 +234,45 @@ private MapSqlParameterSource getLikeParameters(int filmId, int userId) {
         return parameters;
     }
+
+    public Collection<Film> getRecommendations(int userId) {
```
В данном случае более корректно было бы использовать `List`, так как мы работаем с конкретной его реализацией)

YurkinED replied:
ok

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить на данном этапе. Если возникнут вопросы - пишите в слак, буду рад помочь) Хорошего вечера и жду на повторную итерацию🌃

mkudriavtsev committed:
fix: post-review edits

mkudriavtsev committed:
fix: fix minor typos

YurkinED committed:
Merge pull request #23 from YurkinED/kudryavtsev

Правки функционала "Отзывы"

Solntmore committed:
<refactor> В класс ErrorHandler добавлено логирование и хендлер для обработки всех необработанных исключений.

Solntmore committed:
<refactor> В классах FilmDbStorage, FilmService и UserController изменен тип метода getRecommendations с Collection на List.
<refactor> В классе FilmDbStorage приватные методы опущены вниз для соблюдения code style.

Solntmore committed:
<refactor> Интерфейс DirectorInterface переименован в DirectorStorage.
<refactor> В классе FilmDbStorage изменена зависимость. DirectorService заменен на DirectorStorage.

Solntmore committed:
<refactor> В классе DirectorService в методе findDirectorById удалена ошибочная аннотация @PathVariable

Solntmore committed:
<refactor> В классе DirectorService, FilmService, MpaService внедрение произведено через интерфейс хранилища

Solntmore committed:
<refactor> В классе Film BaseEntity заменено на Mpa.

Solntmore committed:
<refactor> В классе FilmController в метод `showMostLikedFilms` добавлена валидация на минимальное значение limit и year, изменено дефолтное значение year с 0 на 10000, в SQL запросе также внесены соответствующие изменения.
<refactor> В классе Film BaseEntity заменено на Genre. Улучшены методы createFilm и updateFilm, теперь имя жанра вносится в объем вместе с id и метод возвращает полноценный объект без значений “null”.
<refactor> В классе FilmDbStorage изменен метод `addGenreAndDirectorToFilm, теперь жанры в Set передаются не по одному, а накапливаются в Set внутри метод, и через сеттер Set жанров передается в фильм.`

IvanSuchilin committed:
add DAO for feed and ref remove-addLike

Solntmore committed:
<feat> В класс ErrorHandler добавлен новый хендлер для обработки ошибок валидации.

IvanSuchilin committed:
added enum etype-operation

MikhailKuznets committed:
feat: добавить валидацию полей в классе User, refactor: сделать внедрение в UserService интерфейcа UserStorage

MikhailKuznets committed:
refactor: изменить Data Type идентификатора id класса BaseEntity с int на long

YurkinED committed:
Merge pull request #27 from YurkinED/solntsev

Solntsev

IvanSuchilin committed:
added enums for reqParametres

IvanSuchilin committed:
Merge branch 'develop' into suchilin

YurkinED committed:
Merge remote-tracking branch 'remotes/origin/kuznetsov' into develop

IvanSuchilin committed:
after merging with develop

YurkinED committed:
Bug fix

YurkinED committed:
Update FilmDbStorage.java

IvanSuchilin committed:
Merge branch 'develop' into suchilin

IvanSuchilin committed:
aftersecon merging with develop

IvanSuchilin committed:
Merge branch 'develop' into suchilin

IvanSuchilin committed:
finished after merging

YurkinED committed:
Merge pull request #28 from YurkinED/suchilin

Suchilin

YurkinED committed:
Final fix

YurkinED committed:
Актуализация схемы

YurkinED committed:
Исправление тестов

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -2,25 +2,30 @@
     private String email;
+
     @NotBlank(message = "Необходимо указать login.")
+    @NotNull(message = "Login не должен быть NULL.")
```
Использование `@NotNull` здесь не имеет смысла, так как проверка на null присутствует в `@NotBlank`

YurkinED replied:
Ок

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/director/DirectorDbStorage.java:
```diff
@@ -0,0 +1,89 @@
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+    @Autowired
+    public DirectorDbStorage(NamedParameterJdbcTemplate namedParameterJdbcTemplate) {
```
Так же можно воспользоваться аннотацией @RequiredArgsConstructor из lombok, благодаря которой автоматически будет сгенерирован конструктор для финальных переменных)

YurkinED replied:
Ок

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -100,6 +95,7 @@ public boolean checkUserExists(int userId) {
         }
     }
 
+    @Override
```
Для работы с друзьями следует использовать другой DAO, так как это отдельная сущность в БД)

YurkinED replied:
Ок

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -110,11 +106,13 @@ public boolean checkFriendshipExists(int userId, int friendId) {
+    @Override
     public Collection<User> showUserFriendsId(int userId) {
         List<User> friendsList = new ArrayList<>();
         SqlRowSet userRows = namedParameterJdbcTemplate.getJdbcTemplate().queryForRowSet(SQL_QUERY_TAKE_FRIENDS_BY_USER_ID, userId);
```
Получение следует произвести в одном запросе, то есть к таблице `FRIENDS` следует сразу присоединить таблицу `USERS`, чтобы не производить n-лишних запросов к БД для получения каждого пользователя отдельно)

YurkinED replied:
Ок

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -124,19 +122,15 @@ public Collection<User> showUserFriendsId(int userId) {
     }
 
+    @Override
     public Collection<User> showCommonFriends(int userId, int friendId) {
```
Получение следует произвести одним запросом, так же фильтрация должна быть осуществлена в нем, чтобы не тратить ресурсы сначала на получение ненужных данных, а затем на их фильтрацию)

YurkinED replied:
Ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserStorage.java:
```diff
@@ -1,5 +1,6 @@
+import ru.yandex.practicum.filmorate.model.Feed;
 import ru.yandex.practicum.filmorate.model.User;
 
 import javax.validation.Valid;
```
@Valid в обоих методах не имеет смысла)

YurkinED replied:
Ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -95,57 +107,155 @@ public void likeFilmOrRemoveLike(int filmId, int userId, boolean flag) {
-        }
-        return film;
+        film.setRating(rs.getInt("rating"));
+        return addGenreAndDirectorToFilm(film);
```
Установку жанров и режиссеров фильмам следует разбить на два отдельных метода, для соблюдения Single Responsibility Principle (Принцип единственной ответственности). Так же следует исправить методы таким образом, чтобы они принимал коллекцию фильмов, чтобы не производить n-лишних запросов к БД, а с помощью условия `IN` мы сначала произведем извлечение всех необходимых жанров и директоров, а только потом будет присоединять их к конкретным фильмам)
Более логично было бы оба метода расположить в своих DAO. И вызывать мы из будем из сервиса фильмов, так как это отдельные сущности в БД)

YurkinED replied:
Ok

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -2,35 +2,42 @@
+@Primary
 public class FilmDbStorage implements FilmStorage {
     private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+    private final DirectorStorage directorStorage;
```
Уровень хранилища не может зависеть от другого хранилища) Хорошим примером является та архитектура, где модуль верхнего уровня зависит от модуля более низкого уровня, но допускается зависимость сервиса от другого сервиса)

YurkinED replied:
Ok

TyutterinYakov reviewed (commented):
Добрый день! Отметил еще некоторые недочеты) Хорошего дня🌅 Если возникнут вопросы - пишите в слак, буду рад помочь!

YurkinED committed:
Bug fix

IvanSuchilin committed:
add friendsDao

IvanSuchilin committed:
Merge branch 'yurkin2' into suchilin

IvanSuchilin committed:
with merge commit yrkin2

YurkinED committed:
Merge pull request #30 from YurkinED/suchilin

Suchilin

YurkinED committed:
Вынес мапперы в отдельный класс

YurkinED committed:
Удаление неиспользуемых импортов, удаление неиспльзуемого UserValidator

YurkinED committed:
Merge pull request #32 from YurkinED/yurkin3

Yurkin3

YurkinED committed:
исправление метода для получения друзей

YurkinED committed:
Update FriendDbStorage.java

MikhailKuznets committed:
refactor: убрать DirectorDbStorage из FilmDbStorage, style: удалить неиспользуемые импорты

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,116 @@
+public class FilmService {
+
+    private final FilmDbStorage filmStorage;
+    private final UserDbStorage userStorage;
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации)

YurkinED replied:
исправили

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,116 @@
+@Service
+public class FilmService {
+
+    private final FilmDbStorage filmStorage;
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации)

YurkinED replied:
Исправивли

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -1,36 +1,51 @@
+@RequiredArgsConstructor
 public class FilmDbStorage implements FilmStorage {
     private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+    private final DirectorStorage directorStorage;
```
Уровень хранилища не может зависеть от другого хранилища) Хорошим примером является та архитектура, где модуль верхнего уровня зависит от модуля более низкого уровня, но допускается зависимость сервиса от другого сервиса)

YurkinED replied:
Испраивили

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -1,36 +1,51 @@
-    @Autowired
-    public FilmDbStorage(NamedParameterJdbcTemplate namedParameterJdbcTemplate) {
-        this.namedParameterJdbcTemplate = namedParameterJdbcTemplate;
+    private final GenreStorage genreStorage;
```
Уровень хранилища не может зависеть от другого хранилища) Хорошим примером является та архитектура, где модуль верхнего уровня зависит от модуля более низкого уровня, но допускается зависимость сервиса от другого сервиса)

YurkinED replied:
Исправили

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -95,57 +98,140 @@ public void likeFilmOrRemoveLike(int filmId, int userId, boolean flag) {
         if (genres.size() > 0) {
-            for (BaseEntity element : film.getGenres()) {
-                int genreId = element.getId();
+            for (Genre element : film.getGenres()) {
```
В данном случае следует использовать `jdbcTemplate.batchUpdate...`, чтобы не совершать n-лишних запросов к БД, а сначала собрать один запрос и отправить его)

YurkinED replied:
Исправили

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -95,57 +98,140 @@ public void likeFilmOrRemoveLike(int filmId, int userId, boolean flag) {
+        namedParameterJdbcTemplate.getJdbcTemplate().update(SQL_QUERY_DELETE_FILMS_DIRECTORS, filmId);
+        Set<Director> directors = film.getDirectors();
+        if (directors.size() > 0) {
+            for (Director element : film.getDirectors()) {
```
В данном случае следует использовать `jdbcTemplate.batchUpdate...`, чтобы не совершать n-лишних запросов к БД, а сначала собрать один запрос и отправить его)

YurkinED replied:
Исправили

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -95,57 +98,140 @@ public void likeFilmOrRemoveLike(int filmId, int userId, boolean flag) {
+
+
+
+    private Film addGenreAndDirectorToFilm(Film film) {
```
Метод следует разбить на два и расположить в DAO жанров и DAO режиссеров) Так же метод должен будет принимать коллекцию фильмов и с помощью SQL условия IN производить получение коллекции, чтобы не производить обращение к БД на каждой итерации цикла

YurkinED replied:
Так DAO не могут между собой общаться. Почему нельзя скрипты выполнить в FilmDBStorage? Обращаться на уровне DAO не рекомендуется между классами же. ИЛи на уровне сервисов обновить логику?

YurkinED replied:
Удалили этот метод

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/friend/FriendDbStorage.java:
```diff
@@ -0,0 +1,69 @@
+
+@Slf4j
+@Repository
+@Primary
```
Использование `@Primary` не является хорошей практикой, хорошо было бы подумать, как этого избежать)

YurkinED replied:
Ушли

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/mpa/MpaDbStorage.java:
```diff
@@ -1,32 +1,29 @@
 
-@Component
+@Repository
+@Primary
```
От `@Primary` так же хорошо было бы уйти)

YurkinED replied:
Ушли

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/ReviewDBStorage.java:
```diff
@@ -0,0 +1,109 @@
+
+@Repository
+@RequiredArgsConstructor
+@Primary
```
Аналогично)

YurkinED replied:
Ушли

TyutterinYakov reviewed (commented):
Добрый день! Отметил оставшиеся недочеты :) Хорошего дня! Если возникнут вопрос - пишите в слак, буду рад помочь!

YurkinED committed:
Merge pull request #33 from YurkinED/kuznetsov3

refactor: убрать DirectorDbStorage из FilmDbStorage, style: удалить н…

YurkinED committed:
Исправление замечаний

Добавил batchUpdate, избавился от цикла добавления Жанров и режиссеров

YurkinED committed:
Избавился от Primary

YurkinED committed:
Merge pull request #34 from YurkinED/yurkin4

Исправление замечаний

TyutterinYakov reviewed (approved):
Добрый день! Вы отлично поработали, и я рад, что Вы уже продвинулись так далеко и не отступили, не бросили, это уже заслуживает огромных похвал! Надеюсь, что работа в команде вам понравилась, и вы остались довольны проделанной работой. В целом - проект выполнен классно. Вы хорошо разобрались с взаимодействием слоев Controller - service - repository и более тесно рассмотрели SQL запросы. Желаю успехов в дальнейшем обучении и продуктивной рабочей недели!🎇🌄 И конечно же немного мотивации))) "Столкнувшись с трудностями, нельзя сдаваться, бежать. Вы должны оценивать ситуацию, искать решения и верить в то, что все делается к лучшему. Терпение – вот ключ к победе!"
